### ISSUE TITLE
Products and attribute values showing incorrectly

### ISSUE THREAD

**Issue by erickmarsh** [Smile-SA/elasticsuite#204]
I have a client using the extension, and we have 2 related problems:
1. We have attributes appearing in the left nav that are not assigned to products in that category.
Example: We are navigating through men's boots, and seeing a style option for "Ballerina". Filtering by this attribute brings us to a product which does not have Ballerina selected for the Style attribute. (it is blank)
http://www.shoebacca.com/catalog/mens-shoes/boots.html?acu_in_style=Ballerina

2. There are products appearing in filtered results that should not be included.
Example: Filtering to find Women's Running Shoes, and "Ballet Flats" appear in the result set
http://www.shoebacca.com/catalog/womens-shoes/athletic-shoes.html?acu_in_sport=Running&p=2 

I verified that none of the associated simple products have any incorrect attribute values which would be causing the improper filtering.

**Comment by erickmarsh** [comment]
Forgot to mention, we are using:
- Smile Extension 2.2.2
- Mage EE 2.1.0
- PHP 7.0.8
- Ubuntu 16.04
- ElasticSearch 2.4.1

**Comment by afoucret** [comment]
It is quite hard for me to figure what happen there since we have never faced this kind of issue. 

I would suggest that the cause of the problem can be one of those 2 (maybe both together) :

1. Did you installed several instances of Magento / ElasticSuite on the same server (let's say production, staging, testing, ...) ? <br/>In this case, is every instance configured to handle it's own ElasticSearch Index as explained in the [wiki](https://github.com/Smile-SA/elasticsuite/wiki/ModuleInstall) ? At least, you must have a differet settings for the `Indices alias name` configuration param, if you don't want your indices to be messed up between the instances ?

2. Did you try to reindex everything at least once through : <br/>
```bin/magento indexer:reindex```

Hope it helps. Thank you for your feedback on this two ones.

**Comment by gewaechshaus** [comment]
@afoucret - sounds like it's kind of related to #195

**Comment by afoucret** [comment]
@gewaechshaus,

I don't think this tow issue are related. @erickmarsh problem looks more like if the index contains product that are not really synced with the database (wrong facets, wrong ids, ...)

We will wait @erickmarsh responses on this ...

**Comment by erickmarsh** [comment]
We are running 3 completely separate environments (Prod, Stage, Dev) with their own web servers, DB and ES so there should be no conflicts. I have reindexed everything again, just to make sure, and the results are the same. 

One note, running the elasticsuite_categories_fulltext returns almost immediately. However, turning on debugging for the extension, it does not appear to be outputting any failures or exceptions (see attached). 
[debug-reindex.txt](https://github.com/Smile-SA/elasticsuite/files/593381/debug-reindex.txt)

**Comment by romainruaud** [comment]
Hello @erickmarsh 

You are right, the index log seems OK.

Do you have any update on this one ? 

Especially, how is looking the "Timberland" boots on your Elasticsearch index ?

**Comment by erickmarsh** [comment]
We have actually logged an issue with Magento as well. Using the built-in Magento ES module we were getting Seg Faults when reindexing the search index. In order to get past the Seg Fault, we updated the /etc/security/limits.conf file to the following

```
* hard nofile unlimited
* soft nofile unlimited
* hard stack unlimited
* soft stack unlimited
```

Here is the output of the index process using the default Magento ES module.

```
ubuntu@ip-10-5-20-209:/var/www/html$ sudo -u www-data bin/magento indexer:reindex catalogsearch_fulltext
Catalog Search indexer process unknown error:
DateTime::__construct(): Failed to parse time string (2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00 2015-11-24 00:00:00) at position 20 (2): Double date specification
ubuntu@ip-10-5-20-209:/var/www/html$
```

I have no idea how the Smile Extension works with the default indexer (if at all), but this is the only piece of additional information I have been able to gather.

**Comment by romainruaud** [comment]
Hello @erickmarsh and thank you for your feedback.

Our extension does nothing with the default indexer, so the data you gave is not relevant to help you with your initial issue.

But, we recently found an issue which could occur when using virtual categories, and having scripts (or custom modules) that loops over categories collection and process "save" on them. This bug can lead to a data loss (emptying the virtual category rule) and therefore result in wrong values and products being shown in the catalog.

Do you use virtual categories ? If yes, do you have any script or module which runs mass operations on categories ? And also, could you check that your virtual categories rules are still present and working properly on your instance ?

Best regards

**Comment by erickmarsh** [comment]
Yes, we do use virtual categories, but the example above is not a virtual category. We do have some custom processes that pull in inventory etc... from an ERP system. We have not seen any issues where a virtual category is emptied or any other issues that are confined to virtual categories.

Thanks!

**Comment by romainruaud** [comment]
Hello @erickmarsh , did you switch back to ElasticSuite finally ? Or are you still using Magento's core implementation ?

Did your "Timberland" boots also appeared as buggy with the Magento's core implementation, or is it appearing only with Elasticsuite enable ?

Can you post here how is looking these Timberland boots in the Elasticsearch index ? To ensure it has no value for the ```acu_in_style``` attribute ?

**Comment by afoucret** [comment]
@erickmarsh,

We have detected a very similar situation in one of our project.
The bug we had was fixed by PR #275 and is not available in the latest releases (2.1.4 and 2.2.6). It is also merged in the upcoming 2.3.x branch and in the master one.

Imho the ticket can be closed by this update. Feel free to reopen if you still have a problem.

**Comment by erickmarsh** [comment]
Thank you! We will give it a try and provide feedback either way.

-Eric

------------------------------
Eric Marsh
eric@ericmarsh.com
www.ericmarsh.com
